### DISCUSSION TITLE
How to pass a database connection to a task

### DISCUSSION THREAD

**Discussion by xorduna** [hibiken/asynq#344] [Q&A]
Hi!

I am starting with Asynq and with Go as a programming language. I have a small project where I need the tasks to have access to a database.

Which would be the proper way to pass a database connection to the context of the task?

I have been looking around, but I am unable to find a way.

Thank you very much!

Xavi

**Comment by xorduna** [comment]
Ok, maybe I am answering my self ...

I found this example in github: https://github.com/IEEE-VIT/cowin-emailer/blob/main/db/setup.go. Where a "global" variable is set. Is this a good Go pattern?

Thanks!

**Reply by gmhafiz** [reply]
**global variable is bad.**

To inject database dependency, several things need to be done.

You need to implement `Handler` interface 
```go
type Handler interface {
	ProcessTask(context.Context, *Task) error
}
```

So, create a new struct and a constructor
```go
type myCustomStruct struct {
    db *sql.DB
}

func NewMethod(db *sql.DB) *myCustomStruct {
    return &myCustomStruct{db: db}
}
```

Once you have these, your method (which is a `myCustomStruct` implementation of `Handler` interface ) will be able to access `sql.DB`

```go
func (p *myCustomStruct) ProcessTask(ctx context.Context, task *asynq.Task) error {
    err := p.db.Query(...
}
```

You only pass the database connection pool once, in `main.go ` to your handler.

```go
db := create connection pool
mux := asynq.NewServeMux()

mux.Handle(task.TypeEmailDelivery, email.NewEmailProcessor(db)
```
